heron/packing/src/java/org/apache/heron/packing/binpacking/TwoDimensionVectorBinPacking.java

```diff
@@ -0,0 +1,379 @@
+ * the org.apache.heron.api.Config.TOPOLOGY_CONTAINER_MAX_CPU_HINT,
+ * org.apache.heron.api.Config.TOPOLOGY_CONTAINER_MAX_RAM_HINT,
+ * org.apache.heron.api.Config.TOPOLOGY_CONTAINER_MAX_DISK_HINT parameters.
+ * If the parameters are not specified then a default value is used for the maximum container
+ * size.
+ * <p>
+ * 3. The user provides a percentage of each container size that will be used for padding
+ * through the org.apache.heron.api.Config.TOPOLOGY_CONTAINER_PADDING_PERCENTAGE
+ * If the parameter is not specified then a default value of 10 is used (10% of the container size)
+ * <p>
+ * 4. The RAM required for one instance is calculated as:
+ * value in org.apache.heron.api.Config.TOPOLOGY_COMPONENT_RAMMAP if exists, otherwise,
+ * the default RAM value for one instance.
+ * <p>
+ * 5. The CPU required for one instance is calculated as the default CPU value for one instance.
+ * <p>
+ * 6. The disk required for one instance is calculated as the default disk value for one instance.
+ * <p>
+ * 7. The RAM required for a container is calculated as:
+ * (RAM for instances in container) + (paddingPercentage * RAM for instances in container)
+ * <p>
+ * 8. The CPU required for a container is calculated as:
+ * (CPU for instances in container) + (paddingPercentage * CPU for instances in container)
+ * <p>
+ * 9. The disk required for a container is calculated as:
+ * (disk for instances in container) + ((paddingPercentage * disk for instances in container)
+ * <p>
+ * 10. The pack() return null if PackingPlan fails to pass the safe check, for instance,
+ * the size of RAM for an instance is less than the minimal required value.
+ */
+public class TwoDimensionVectorBinPacking extends AbstractPacking {
+
+  private static final Logger LOG = Logger.getLogger(TwoDimensionVectorBinPacking.class.getName());
+
+  private int numContainers = 0;
+
+  private PackingPlanBuilder newPackingPlanBuilder(PackingPlan existingPackingPlan) {
+    return new PackingPlanBuilder(topology.getId(), existingPackingPlan)
+        .setDefaultInstanceResource(defaultInstanceResources)
+        .setMaxContainerResource(maxContainerResources)
+        .setRequestedContainerPadding(padding)
+        .setRequestedComponentResource(componentResourceMap)
+        .setInstanceConstraints(Arrays.asList(new MinRamConstraint(), new MinCpuConstraint()))
+        .setPackingConstraints(Arrays.asList(new ResourceConstraint(),
+            new InstanceDensityConstraint(maxNumInstancesPerContainer)));
+  }
+
+  /**
+   * Get a packing plan using First Fit Decreasing
+   *
+   * @return packing plan
+   */
+  @Override
+  public PackingPlan pack() {
+    PackingPlanBuilder ramFirstPlanBuilder = newPackingPlanBuilder(null);
+    PackingPlanBuilder cpuFirstPlanBuilder = newPackingPlanBuilder(null);
+
+    // Get the instances using FFD allocation
+    try {
+      numContainers = 0;
```

Comment:
Some comments might be helpful here. I can imaging me trying to delete the second "numContainers = 0;" since it looks like a bug.